### ISSUE TITLE
[typescript-angular] authorization header not send in 2.3.0

### ISSUE THREAD

**Issue by meisterlampe** [swagger-api/swagger-codegen#7342]
##### Description

In my project I use the following config for basic auth:
```
{
        accessToken: undefined,
        apiKeys: undefined,
        password: "pass",
        username: "user",
        withCredentials: false,
}
```

The problem showed up after upgrading from swagger-codegen 2.2.3 to 2.3: Requests are send without the authorization header. Instead properties like `lazyupdate` are contained. If I check the generated http headers object in chromeDevTools I indeed see all these properties as part of the headers object and `lazyupdate` does contain my authorization-header with the base64 string:
![screenshot_2018-01-08_14-28-51](https://user-images.githubusercontent.com/16793613/34672899-931633ce-f480-11e7-9f44-5be8a98c8da0.png)

Here is a screenshot showing the request when I use the default typescript-angular generator for the default ngVersion (4.3):
![missingheaders](https://user-images.githubusercontent.com/16793613/34672702-c7ad3304-f47f-11e7-8970-7e8f9850deeb.png)

In codegen 2.2.3 (or if I pass in --additional-properties=ngVersion=2) everything works fine.
This is the expexted result:
![ngversion2](https://user-images.githubusercontent.com/16793613/34672767-05957bb8-f480-11e7-9246-68e877a88b3c.png)

Do you have any idea, what I have to do to successfully upgrade to the angular template using httpClient? Is this maybe a bug?

Thanks in advance :)

##### Swagger-codegen version

2.3.0

**Comment by kenisteward** [comment]
Make sure you are properly injecting the Configuration object.

In this case you can simply provide a Configuration object that you can change whenever you get the user information.

https://github.com/swagger-api/swagger-codegen/tree/master/samples/client/petstore/typescript-angular-v4.3/npm

Let me know if you need any more information.

**Comment by kenisteward** [comment]
Also where are lazyinit and lazy update coming from?  It's possible your proprietary code is not compatible with these things.

**Comment by kenisteward** [comment]
Also it looks like it's try to users headers as the headers values instead of the internal .headers which is why those values are sent like that.  I'm looking into this now.

**Comment by kenisteward** [comment]
@meisterlampe  Make sure if you are setting service.defaultHeaders that you are doing

```
service.defaultHeaders = service.defaultHeaders.set('Authorization', 'fqwefqwefqwefqewfqwefqfe');
```

and not 

```
service.defaultHeaders.set('Authorization', 'asdfqwefqwefqewfqewfqewf');
```

headers are immutable as of Angular 4.3.  You should also be using headers from @angular/common/http instead of the other module.

My other question is could you possible show me where you are setting the config values?

**Comment by meisterlampe** [comment]
Hi Keni,

thank you very much for your fast answers.

I'm using Ionic 3.19 and angular 5.0.0. 
As said before, using the generator with ngVersion=2 is working fine (this uses the old http-Object and not the httpClient like ngVersion=4.3

I'm passing the configuration, when creating the Service. The configuration is never changing during runtime:
```
let config = new Configuration({
  accessToken: undefined,
  apiKeys: undefined,
  password: credentials.password,
  username: credentials.user,
  withCredentials: false,
});
return new DefaultService(this.httpClient, baseUrl, config);
```

The first screenshot above is actually the generated default.service-class. The configuration values are correctly showing up in the following Screenshot (User:"user", Password: "superSecretPass"):
![screenshot_2018-01-09_08-20-13](https://user-images.githubusercontent.com/16793613/34709358-34cbd8c2-f517-11e7-8c1f-026c7fd482d7.png)

The generated code is then correctly setting the headers the way you suggested in line 755:
`headers = headers.set('Authorization', 'Basic ' + btoa(this.configuration.username + ':' + this.configuration.password));`

The problem is now, that after executing `headers.set` the headers-object does contain the authorization header in `lazyupdate`. `lazyupdate` must be something used by the headers-class. Everything you see is the generated code without any modifications.
I also took a screenshot of that. Note that the code execution pointer is in line 759. The headers.set command has just been executed here (and no other code has run).
![screenshot_2018-01-09_08-35-19](https://user-images.githubusercontent.com/16793613/34709603-26e15902-f518-11e7-99b8-0ce2f6a60380.png)


#### Summary: 
Line 755 `headers = headers.set('HEADER', 'VALUE'));` results in a new `HttpHeaders`-object containing `headers.lazyupdate[0] == { name: 'HEADER', value: 'VALUE'}`

**Comment by kenisteward** [comment]
It turns out this is to do with the way we should be handling headers.  It used to be (with mutable headers) this would work perfectly fine.  Now with immutable headers we need to take more care to make sure we tell the headers to process into normal headers (by getting it to call the private init method)

We can fix this on our side but I may open a ticket with Angular.  I feel like the implementation right now (making sure to call set again) should work perfectly fine but it happens that this.init is never called.

**Comment by kenisteward** [comment]
Is it possible for you to setup a www.stackblitz.com with your example so I can further debug yours in particular?

**Comment by kenisteward** [comment]
@meisterlampe what version of Angular are you using?  Just wondering because you want the client lib to match the same lib as the one you are using.  e.g. for my apps I generate 5.1.x

**Comment by kenisteward** [comment]
I've found the issue!!

https://github.com/angular/angular/blob/4.3.0/packages/common/http/src/client.ts#L330-L337

this is the code at 4.3.0 which is probably the version of angular you have set to use right now

in the beta of 5 this becomes

https://github.com/angular/angular/blob/5.0.0-beta.7/packages/common/http/src/client.ts#L354-L360

which I've tested properly propagates the header information (in a plunker)  could you try putting a more recent angular version for your main one you have?

This also explains why the internal things were being sent instead the actual headers.

**Comment by meisterlampe** [comment]
Wow thank you for your hints. This actually got me to the root of the problem.
First of all I made sure, I'm really using angular 5.0.0, but that was not the only issue.
Thanks to your last post, I started debugging all the way down to the point, where the header is built.
There I noticed, that this was completely different code... wait... is this @angular/http? :)

I forgot to remove the old http provider completely and somehow the generated swagger-service was using this ...even though I created the DefaultService by passing the HttpClient.

Thank you very very much, for resolving my issue. I'm very thankful for that!
